cpp/tests/io/fst/quote_normalization_test.cu

```diff
@@ -0,0 +1,328 @@
+  NUM_SYMBOL_GROUPS   ///< Total number of symbol groups
+};
+// Aliases for readability of the transition table
+constexpr auto TT_OOS            = dfa_states::TT_OOS;
+constexpr auto TT_DQS            = dfa_states::TT_DQS;
+constexpr auto TT_SQS            = dfa_states::TT_SQS;
+constexpr auto TT_DEC            = dfa_states::TT_DEC;
+constexpr auto TT_SEC            = dfa_states::TT_SEC;
+constexpr auto TT_NUM_STATES     = static_cast<char>(dfa_states::TT_NUM_STATES);
+constexpr auto NUM_SYMBOL_GROUPS = static_cast<uint32_t>(dfa_symbol_group_id::NUM_SYMBOL_GROUPS);
+// The i-th string representing all the characters of a symbol group
+// TODO: We should reinstantiate this lookup table approach once
+// https://github.com/rapidsai/cudf/pull/14561 is merged std::array<std::vector<SymbolT>,
+// NUM_SYMBOL_GROUPS - 1> const qna_sgs{
+//  {{'\"'}, {'\''}, {'\\'}, {'\n'}}};
+// Temporary workaround:
+struct SymbolToSymbolGroup {
+  CUDF_HOST_DEVICE uint32_t operator()(SymbolT symbol) const
+  {
+    switch (symbol) {
+      case '\"': return static_cast<uint32_t>(dfa_symbol_group_id::DOUBLE_QUOTE_CHAR);
+      case '\'': return static_cast<uint32_t>(dfa_symbol_group_id::SINGLE_QUOTE_CHAR);
+      case '\\': return static_cast<uint32_t>(dfa_symbol_group_id::ESCAPE_CHAR);
+      case '\n': return static_cast<uint32_t>(dfa_symbol_group_id::NEWLINE_CHAR);
+      default: return static_cast<uint32_t>(dfa_symbol_group_id::OTHER_SYMBOLS);
+    }
+  };
+};
+// Transition table
+std::array<std::array<dfa_states, NUM_SYMBOL_GROUPS>, TT_NUM_STATES> const qna_state_tt{{
+  /* IN_STATE      "       '       \       \n    OTHER  */
+  /* TT_OOS */ {{TT_DQS, TT_SQS, TT_OOS, TT_OOS, TT_OOS}},
+  /* TT_DQS */ {{TT_OOS, TT_DQS, TT_DEC, TT_OOS, TT_DQS}},
+  /* TT_SQS */ {{TT_SQS, TT_OOS, TT_SEC, TT_OOS, TT_SQS}},
+  /* TT_DEC */ {{TT_DQS, TT_DQS, TT_DQS, TT_OOS, TT_DQS}},
+  /* TT_SEC */ {{TT_SQS, TT_SQS, TT_SQS, TT_OOS, TT_SQS}},
+}};
+// The DFA's starting state
+constexpr char start_state = static_cast<char>(TT_OOS);
+struct TransduceToNormalizedQuotes {
+  /**
+   * @brief Returns the <relative_offset>-th output symbol on the transition (state_id, match_id).
+   */
+  template <typename StateT, typename SymbolGroupT, typename RelativeOffsetT, typename SymbolT>
+  constexpr CUDF_HOST_DEVICE SymbolT operator()(StateT const state_id,
+                                                SymbolGroupT const match_id,
+                                                RelativeOffsetT const relative_offset,
+                                                SymbolT const read_symbol) const
+  {
+    // -------- TRANSLATION TABLE ------------
+    // state | read_symbol -> output_symbols
+    // DQS   | *           -> *
+    // DEC   | *           -> *
+    // OOS   | '           -> "
+    // OOS   | *           -> *
+    // SQS   | "           -> \"
+    // SQS   | '           -> "
+    // SQS   | *           -> *
+    // SEC   | *           -> *
```

Comment:
My idea was that this would reflect _all_ transitions, and the below `SPECIAL CASES` would exclude the `<state>, * -> *` cases from this table. 

Currently this part of the table is not representative of the latest changes, where we now have `SQS | \\ -> <nop>`. Should we update this part of the table or just get rid of this part of the table? Do you think the special cases part is sufficient for future maintainers?

Reply:
I think we should retain the non-special transitions for completeness - we can construct the DFA by just reading the transition table. How about dividing the table into non-special and special transitions? We can also make the input symbols used explicit. For instance, `OOS | <s> \in Sigma\{'} -> <s>` in non-special and `OOS | ' -> "` in special.